### ISSUE TITLE
ASP.NET MVC crashing IIS with Event 5011 in Event Viewer

### ISSUE THREAD

**Issue by bnemetchek** [agracio/edge-js#17]
Running the supplied example:

```
var func = Edge.Func(@"
                    return function (data, callback) {
                        callback(null, 'Node.js welcomes ' + data);
                    }
                "); // crashes here

```
using version 9.3.2/64-bit. VS2017.

A process serving application pool 'DefaultAppPool' suffered a fatal communication error with the Windows Process Activation Service. The process id was '6448'. The data field contains the error number.

Been trying to get this working for some time.  If anyone has any advice, I would appreciate.

thanks!

**Comment by agracio** [comment]
It hard to figure out what the problem is without actual exception stack trace.
Have you tried running the code in a simple console app to see if it works outside IIS?
How are you importing Edge.Js libraries, NuGet package? Which one you using?

**Comment by bnemetchek** [comment]
I'm using the Nuget package:
https://www.nuget.org/packages/EdgeJs/9.3.2.
There doesn't seem to be a stack trace to be had.  The process just terminates.  It would likely work in a console app as I believe it is a problem with running it on IIS/WIN10.  

I'm using .Net 4.6.  Is that supported?

**Comment by agracio** [comment]
Could you create an sample repo and I will take a look at it.

**Comment by scottned** [comment]
I've been trying to resolve this problem for a long time as well while trying to upgrade from Edge 6.11.3.  I'm using the 9.3.2 NuGet package and when I try to invoke Edge.Func the whole process just exits.  I've looked at the output in procmon and it looks like edge/x64/node.dll is loading and then looking at all the packages in my node_modules directory and then nothing else happens.  Presumably the process terminated after that.

I have node 8.9.3 installed locally (used [Chocolatey](https://chocolatey.org/) to install)

I also attached the Debug Diagnostic Tool to IIS and got this stack trace:
```
ntdll!NtTerminateProcess 
ntdll!RtlExitUserProcess+5b 
kernel32!ExitProcessImplementation+a 
mscoreei!RuntimeDesc::ShutdownAllActiveRuntimes+287 
mscoree!ShellShim_CorExitProcess+11e 
ucrtbase!try_cor_exit_process+3b77d 
ucrtbase!exit_or_terminate_process+64 
ucrtbase!common_exit+6f 
node!node::DecodeWrite+198a 
node!v8::internal::RegisterConfiguration::AreAliases+1e65 
node!v8::internal::RegisterConfiguration::AreAliases+1586 
node!v8::internal::RegisterConfiguration::AreAliases+527 
node!v8::internal::RegisterConfiguration::AreAliases+40b 
0x000003af`9a6047a1 
0x000000fd`00000000 
.... a bunch of other similar lines
0x000000d0`e797cf20 
0x000003af`9a604161 
DomainBoundILStubClass.IL_STUB_PInvoke(Int32, System.String[])+116 
[[InlinedCallFrame] (EdgeJs.Edge.NodeStartx64)] EdgeJs.Edge.NodeStartx64(Int32, System.String[]) 
EdgeJs.Edge+<>c__DisplayClass11_0.<Func>b__0()+105 
mscorlib_ni!System.Threading.ExecutionContext.RunInternal(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)+160 
mscorlib_ni!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object, Boolean)+14 
mscorlib_ni!System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext, System.Threading.ContextCallback, System.Object)+52 
mscorlib_ni!System.Threading.ThreadHelper.ThreadStart()+52 
[[GCFrame]] 
[[DebuggerU2MCatchHandlerFrame]] 
[[ContextTransitionFrame]] 
[[DebuggerU2MCatchHandlerFrame]] 
```
I'm a bit stumped on what to try next.

**Comment by peter-sabath** [comment]
Have your checked your IIS config & Node config? 

Event 5011 is not an application crash event

[https://technet.microsoft.com/en-us/library/cc735271(v=ws.10).aspx](url) lists some condition and checks for this event.

Is the given process ID the one of IIS?
Do you see any other "AppCrash" events in the log near the 5011 event?

Looking at the stack trace it seems that the "process" is terminated gracefully and no exceptional crash condition like an access violation has occurred.

Since "DecodeWrite" called from "RegisterConfiguration" causes the fatal exit, it is possible that there is a buffer overrun with a string.

**Comment by agracio** [comment]
So if I understand this correctly if you use original Edge.Js you get AccessViolation exception while if you use this fork the process just quits quietly? This was also reported on another issue: https://github.com/tjanczuk/edge/issues/611.
I have merged https://github.com/tjanczuk/edge/pull/573 into this repo to address AccessViolation exception and was under the impression that it did solve the problem, but apparently it only works on some systems and not others. 
If anyone could create a very simple repo that reproduces the problem I will try to get to the bottom of it ASAP.
Please include instructions on the repo for running the project.

**Comment by bnemetchek** [comment]
I will put one together tomorrow.  If I zip up the solution, how best to send?

**Comment by agracio** [comment]
Create a Github repo.

**Comment by bnemetchek** [comment]
Repo at [https://github.com/bnemetchek/EdgeJsIssue17Sample](https://github.com/bnemetchek/EdgeJsIssue17Sample)

**Comment by agracio** [comment]
So far I was not able to resolve the issue. 
The exception occurs here: https://github.com/agracio/edge-js/blob/master/src/double/Edge.js/dotnet/EdgeJs.cs#L54 and in .NET 4.0+ it is expected behaviour for unmanaged code exception to terminate the process rather than handle exception as per https://stackoverflow.com/a/5793564. 
I will continue investigation into the root cause of the problem.
@bnemetchek I think you said previously that you were having trouble creating sample project that would replicate the problem as it was working at first and you had to adjust project settings to make it fail? What kind of changes did you have to make?

**Comment by bnemetchek** [comment]
I had to use "Local IIS" instead of "IIS Express" on the web project properties page.  That was the last thing I had to change.  The other changes involved adding security stuff into the app but that is all in the repro so you should be able to make it fail.  I'm using latest Win10/x64 and VS2017.

**Comment by agracio** [comment]
No luck so far, I cannot get it to throw an exception on IIS/IIS Express even after changing machine.config. After I installed IIS on my machine I can get the project to fail on IIS Express as well. Created a new project from scratch and it still happens on IIS Express.

**Comment by IntranetFactory** [comment]
We today had Event 5011 A process serving application pool 'XXX' suffered a fatal communication error with the Windows Process Activation Service with Node.Js v8.9.3

The problem happened only on Windows 2012 R2 and was related to a missing api-ms-win-crt-runtime-l1-1-0.dll. Fixing that required to fix a broken VS 2015 Redist, which required to install a missing OS patch. I can provide the tedious details in the case your problem is also related to Windows 2012 R2.

**Comment by agracio** [comment]
I have the issue on Win10, but will check for missing dll. Was the problem apparent from event log or did you have to figure out why it happens?

**Comment by agracio** [comment]
Actually looking at event log on my Win10 machine it throws `System.AccessViolationException`, so that explains it.

**Comment by agracio** [comment]
@bnemetchek could you check IIS Express for me again. On my machine it works once, but if you click the button again it crashes with `System.AccessViolationException` in the event log.

**Comment by IntranetFactory** [comment]
We found another reason causing an Event 5011 of w3wp.exe: when we use v6 debug option with v8.9.3
` Environment.SetEnvironmentVariable("EDGE_NODE_PARAMS", "--debug=5858")` 
double_edge.js is not even loaded. So `nodeStart` seems to fail quite early.

when using 
` Environment.SetEnvironmentVariable("EDGE_NODE_PARAMS", "--inspect=127.0.0.1:9229")`

it works.

**Comment by agracio** [comment]
I had suspicion that it does not load _double_edge.js_ but could not prove 100%. I did not have time to look into it this week, but hoping to make progress in a few days. @IntranetFactory your comments are really helpful.

**Comment by paulomunoz** [comment]
> We today had Event 5011 A process serving application pool 'XXX' suffered a fatal communication error with the Windows Process Activation Service with Node.Js v8.9.3
> The problem happened only on Windows 2012 R2 and was related to a missing api-ms-win-crt-runtime-l1-1-0.dll. Fixing that required to fix a broken VS 2015 Redist, which required to install a missing OS patch. I can provide the tedious details in the case your problem is also related to Windows 2012 R2.
@IntranetFactory I am having an issue (which I believe to be the same), on Windows 2012 R2. Can you please provide me with the tedious details?

**Comment by paulomunoz** [comment]
I have more info that may help. I created a console application, ultra simple, and found the following:
- Edge.Js 7.10.1 -> Works fine
- Edge.Js 8.2.1+ -> Closes silently!

**Comment by IntranetFactory** [comment]
@paulomunoz we found some systems where installing VS 2015 Redist wasn't sufficient to run v8.9.3. Event after the installation of VS 2015 Redist  api-ms-win-crt-runtime-l1-1-0.dll and other api-ms-*.dll files still had been missing.

We finally came to the conclusion that KB2919355 had been missing on these systems. The installation of KB2919355 before VS 2015 Redist solved the deployment issues to Windows 2012 R2 for us.

**Comment by agracio** [comment]
> I have more info that may help. I created a console application, ultra simple, and found the following:
> Edge.Js 7.10.1 -> Works fine
> Edge.Js 8.2.1+ -> Closes silently!

Can you create a Github repo.

**Comment by paulomunoz** [comment]
> Can you create a Github repo.

https://github.com/paulomunoz/edge-js-test

**Comment by LucidObscurity** [comment]
Just a heads up. Ran into issues with IIS as well, but with Microsoft.AspnetCore, IIS is just a proxy. Anyway ran the app standalone/kestrel via cmd and got this error: ... api-ms-win-crt-runtime-l1-1-0.dll is missing ...
Since this machine (server 20018 R2 SP1) already had two versions of VC++ 2015 installed I went with VC++ 2017 since my dev box (Win 10 v1703) doesn't have 2015 but does have 2017 (from here under 'Other Tools and Frameworks': ([https://visualstudio.microsoft.com/downloads/])). This got me past that particular error...but I've yet to try back under IIS...will update with results.
